tv_apps/tv-deck/js/tv_deck.js

```diff
@@ -0,0 +1,538 @@
+                        document.getElementById('pin-button-contextmenu');
+    this.menuButton = document.getElementById('menu-button');
+    this.loadingIcon = document.getElementById('loading-icon');
+    this.bubbleElement = document.getElementById('bubble-animation');
+
+    // Other dependencies
+    this.keyNavigatorAdapter = new KeyNavigationAdapter();
+    this.keyNavigatorAdapter.init();
+    this.keyNavigatorAdapter.on('move', this._onSwitch.bind(this));
+    this.keyNavigatorAdapter.on('enter-keyup', this._onEnter.bind(this));
+
+    this.simpleKeyNavigation = new SimpleKeyNavigation();
+    this.simpleKeyNavigation.start(
+      [this.pinButton, this.menuButton],
+      SimpleKeyNavigation.DIRECTION.HORIZONTAL
+    );
+
+    // Event listeners setup
+    window.onhashchange = this._onHashChange.bind(this);
+    document.addEventListener('keyup', this._onEnterNumber.bind(this));
+
+    // Flash animation
+    this.channelPanel.addEventListener('animationend', function(){
+      this.channelPanel.classList.remove('flash');
+    }.bind(this));
+
+    // Hide loadingIcon when animationend
+    this.loadingIcon.addEventListener('animationend', function(){
+      this.loadingIcon.classList.add('hidden');
+    }.bind(this));
+
+    // Initialize TVDeck. Fetch initial (tuner, source, channel) values from
+    // either URL hash or localStorage, and then scan all available tuners,
+    // sources and channels.
+    this.fetchSettingFromHash();
+    this.scanTuners();
+  };
+
+  var proto = {};
+
+  // Current tuner object playing by TV (we may have another tuner for recording
+  // in the future).
+  Object.defineProperty(proto, 'playingTuner', {
+    get: function() {
+      return this.currentTuners[this.playingTunerId];
+    }
+  });
+
+  // Current source object playing by TV (we may have another source for
+  // recording in the future).
+  Object.defineProperty(proto, 'playingSource', {
+    get: function() {
+      return this.playingTuner.sources[this.playingSourceType];
+    }
+  });
+
+  // Current channel object playing by TV (we may have another source for
+  // recording in the future).
+  Object.defineProperty(proto, 'playingChannel', {
+    get: function() {
+      var playingSource = this.playingSource;
+      var index = playingSource.channelIndexHash[this.playingChannelNumber];
+      return playingSource.channels[index];
+    }
+  });
+
+  // Current hash
+  Object.defineProperty(proto, 'currentHash', {
+    get: function() {
+      return this.playingTunerId + ',' + this.playingSourceType +
+             ',' + this.playingChannelNumber;
+    }
+  });
+
+  /**
+   * Fetch tuner, source and channel settings from URL hash. If URL hash is
+   * empty, then fetch settings from localStorage.
+   */
+  proto.fetchSettingFromHash = function td_fetchSettingFromHash() {
+    var hash;
+    hash = window.location.hash.substring(1);
+    if (hash.length === 0) {
+      hash = window.localStorage.getItem('TV_Hash');
```

Comment:
Please use asyncStorage instead of localStorage. The performance of localStorage is worse.